### ISSUE TITLE
[Proposal] [Documentation] Improve the low show rate caused by adaptive_banner sample code

### ISSUE THREAD

**Issue by mulderpf** [googleads/googleads-mobile-flutter#780]
## Plugin Version

Version 2.1.0

## Details

EDIT: I am raising this issue as I have updated my app from firebase_admob: ^0.10.3 to Google Mobile Ads.  With the former, I used to achieve a show rate of around 96%, with these delays, my show rate has now dropped to around 35%.  This is not ideal at all as it means that most users don't get to even see the ads and it is impacting my revenue.

1. I noticed some really slow load times of banner ads in my production app.  The fill requests are typically completed within around 400ms, but it takes about 5 times as long for the banner to actually show up.  Taking 2 seconds to display an ad in production is not ideal, especially if the request took around 300ms to complete.  If a bid request was fulfilled within 300ms, I would expect the ad to be shown ideally within no more than 1 second.  
2. I took the Adaptive_Banner_Example and just added a few basic pieces of logging to strip away any potential complexity I might have added (I tested with my own ads and also the test ones with no real difference in performance):

Added this in the top:
```
class AdaptiveBannerExampleState extends State<AdaptiveBannerExample> {
  BannerAd? _bannerAd;
  bool _isLoaded = false;
  late DateTime _start;
  final String _adUnitId = Platform.isAndroid
  ... rest of code as per example
```
And then I updated _loadAd() with the following:
```
  void _loadAd() async {
    // Get an AnchoredAdaptiveBannerAdSize before loading the ad.
    _start = DateTime.now();
    final size = await AdSize.getCurrentOrientationAnchoredAdaptiveBannerAdSize(
        MediaQuery.of(context).size.width.truncate());

    if (size == null) {
      // Unable to get width of anchored banner.
      return;
    }

    BannerAd(
      adUnitId: _adUnitId,
      request: const AdRequest(),
      size: size,
      listener: BannerAdListener(
        // Called when an ad is successfully received.
        onAdLoaded: (ad) {
          setState(() {
            _bannerAd = ad as BannerAd;
            _isLoaded = true;
            DateTime end = DateTime.now();
            print(
                'ADTEST: Ad took ${end.difference(_start).inMilliseconds}ms to load');
          });
        },
        // Called when an ad request failed.
        onAdFailedToLoad: (ad, err) {
          DateTime end = DateTime.now();
          print(
              'ADTEST: Ad failed to load after ${end.difference(_start).inMilliseconds}ms. ${err.toString()}');
          ad.dispose();
        },
        // Called when an ad opens an overlay that covers the screen.
        onAdOpened: (Ad ad) {
          print('ADTEST: Ad opened');
        },

        // Called when an ad removes an overlay that covers the screen.
        onAdClosed: (Ad ad) {
          print('Ad closed');
        },
        // Called when an impression occurs on the ad.
        onAdImpression: (Ad ad) {
          DateTime end = DateTime.now();
          print(
              'ADTEST: Ad impression after ${end.difference(_start).inMilliseconds}ms');
        },
      ),
    ).load();
  }

```

4. I switched to flutter master channel and also ran this in release mode, and the slowness still shows up.

**Target Platform: Testing this on Android**
**Target OS version: Tested with Android 11 and Android 13**
**Devices:Emulator with Android 13, Samsung Galaxy S20 with Android 11 and Pixel 6 with Android 13**

## Logs

<details>
<summary>Logs</summary>

Logs from the application filtered with my added logging:
```
I/flutter (  321): ADTEST: Ad took 1850ms to load
I/flutter (  321): ADTEST: Ad impression after 2497ms
```

The ad inspector tells me that bidding took 459ms to fulfill.

```
Analyzing adaptive_banner_example...                                    

   info • Don't invoke 'print' in production code • lib/main.dart:105:13 • avoid_print
   info • Don't invoke 'print' in production code • lib/main.dart:112:11 • avoid_print
   info • Don't invoke 'print' in production code • lib/main.dart:118:11 • avoid_print
   info • Don't invoke 'print' in production code • lib/main.dart:123:11 • avoid_print
   info • Don't invoke 'print' in production code • lib/main.dart:128:11 • avoid_print
```

```
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel master, 3.8.0-14.0.pre.25, on macOS 13.2.1 22D68 darwin-x64, locale en-GB)
[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.1)
[✓] Xcode - develop for iOS and macOS (Xcode 14.2)
[✓] Android Studio (version 2021.2)
[✓] VS Code (version 1.75.1)
[✓] Connected device (3 available)
[✓] Network resources

• No issues found!
```

</details>

**Comment by huycozy** [comment]
@mulderpf Thanks for the detailed report. But I think this is depending on factors (internet speed, ads server response time, account limitation status, etc) that I'm not sure it's related to plugin code. 

For a more thorough investigation on this, can you try checking this with the [native sample code](https://github.com/googleads/googleads-mobile-android-examples)?

**Comment by mulderpf** [comment]
As much as I would like to spend more time trying to get to the bottom of this, I have little recent experience of running native code, I'm only set up for Flutter development.

I do not have much more sophisticated tools to see where the missing 1.5seconds are spent, but it is pretty obvious in the most basic example, run on multiple devices that there is a delay which needs investigation.  The only reason I started looking into this was because I was seeing a huge drop in the show rate after releasing my app to production yesterday (which now uses the new package).  Given a fairly big user base (around 8500 DAU), this should even out things like internet speed and availability, ads server response times etc - it's the same admob account, same user base, but significantly different performance from a previous version to the current.

In fact, I am not even sure what to do if I find something in the native sample code - it bears no resemblance to anything that I can implement in my Flutter app, except maybe for a huge learning curve (which for me is why I'm using Flutter, so that I don't have to worry about native code).

Surely if you are able to replicate this same delay (which is easy to do), it might be worth investigating with better tools than I have?

I am not trying to be difficult, but I'm raising this with the teams who should know how to investigate this deeper since I'm not the expert with this.  Being asked to investigate on my own with something I have little experience with is a bit counter-intuitive as I'm likely to spend weeks trying to get the thing to run and then have no idea how to actually troubleshoot, nevermind get it into my Flutter project.

Feel free to close this then as I am just not equipped to provide the feedback that you want right now.

**Comment by mulderpf** [comment]
Some further feedback on this.  Instead of loading the ad widget on every page, I've created a shell UI with a BottomNavigationBar which keeps the ad loaded while the user navigates through the rest of the app (this mimics the exact same behaviour as the previous AdMob plugin (firebase_admob).

The previous plugin gives a show rate of around 95%-98%; the new one with the banner ad loaded at the bottom is still just below 40%.  When testing, I have double checked that the ad widget is only created once, so I am still not completely sure why the show rate is so low.  I started thinking it was with my implementation., but it doesn't appear to be the case.

**Comment by mulderpf** [comment]
I have another update and I figured out the issue with the low show rate (turns out that the root cause of the low show rate is not what I thought it was - it's not down to the load times, but rather a bunch of ads being loaded unnecessary because of a bug in the example.  The issue is with the Adaptive Banner example that I was using - specifically that it calls _loadAd() in "didChangeDependencies".  What I found in my production app, is that this is called each time when the keyboard is used, thus resulting in a reload of the ad with each keystroke!  This is causing a bunch of ads to be loaded, but never shown to the user.

Be careful in using the example as I consider this a bug.  Don't reload the ad in didChangeDependencies or at least do another check before just reloading the ad in this method.

**Comment by huycozy** [comment]
Thanks for the response. 

I tried running your code with log and also tried adding logs to native android, the results of 3 attempts per example are as below:

Based on the printed logs, I can be seen that the time difference on Flutter is larger than native but not too obvious (>10ms from loading to impression). The native Android side is faster (only 1-2ms difference).

- Flutter ([flutter adaptive_banner_example](https://github.com/googleads/googleads-mobile-flutter/tree/main/samples/admob/adaptive_banner_example/) with additional log as above [comment](https://github.com/googleads/googleads-mobile-flutter/issues/780#issue-1594973868))

```
I/flutter (21109): ADTEST: Ad took 4623ms to load
I/flutter (21109): ADTEST: Ad impression after 4640ms
```

```
I/flutter (21109): ADTEST: Ad took 2579ms to load
I/flutter (21109): ADTEST: Ad impression after 2591ms
```

```
I/flutter (21109): ADTEST: Ad took 2616ms to load
I/flutter (21109): ADTEST: Ad impression after 2628ms
```

- Android Native Adaptive Banner ([source code demo](https://github.com/huycozy/googleads-mobile-android-examples/blob/flutter_ads_request_issue_780/kotlin/admob/AdaptiveBannerExample/app/src/main/java/com/google/android/gms/example/adaptivebannerexample/MyActivity.kt))

```
D/MyActivity: onAdImpression: ADTEST: -2561ms
D/MyActivity: onAdLoaded: ADTEST: -2562ms
```

```
D/MyActivity: onAdImpression: ADTEST: -2457ms
D/MyActivity: onAdLoaded: ADTEST: -2459ms
```

```
D/MyActivity: onAdLoaded: ADTEST: -2499ms
D/MyActivity: onAdImpression: ADTEST: -2501ms
```

Based on your [above comment](https://github.com/googleads/googleads-mobile-flutter/issues/780#issuecomment-1449532089), the cause of this issue is the implementation in your source code. So, I will close the issue from here as solved. If you disagree, please write in the comments and I will reopen it.

**Comment by mulderpf** [comment]
I do believe that the example should be updated as there are many people who use this as a base and you will see that Flutter is being blamed for a really low show rate.  I have updated my app in production and my show rate is back to normal (above 90%).

The code I changed from the example to quickly fix this is:
`  @override
  void didChangeDependencies() {
    super.didChangeDependencies();
    Orientation newOrientation = MediaQuery.of(context).orientation;
    if (!_isLoaded || _currentOrientation != newOrientation) {
      _currentOrientation = newOrientation;
      _loadAd();
    }
  }
`

I do believe that the examples should be good enough to replicate in production.  In the current state, I think a small fix would help.  Feel free to ignore this and re-close as I will link other pages to this one where people are struggling with low show rates.
